### PR TITLE
Cuda Graph implementation of Kokkos Graphs

### PR DESCRIPTION
**Pull request by dhollman**
The remainder of the Cuda Graph implementations of Kokkos Graphs as well as tests to verify that it works

### CODE REVIEW

File: cmake/kokkos_compiler_id.cmake [hunk 1]
```diff
@@ -92,27 +92,27 @@ ENDIF()
...
-SET(KOKKOS_MESSAGE_TEXT "${KOKKOS_MESSAGE_TEXT}\n    GCC        4.8.4 or higher")
-SET(KOKKOS_MESSAGE_TEXT "${KOKKOS_MESSAGE_TEXT}\n    Intel     15.0.2 or higher")
-SET(KOKKOS_MESSAGE_TEXT "${KOKKOS_MESSAGE_TEXT}\n    NVCC      9.0.69 or higher")
+SET(KOKKOS_MESSAGE_TEXT "${KOKKOS_MESSAGE_TEXT}\n    Clang      4.0.0 or higher")
+SET(KOKKOS_MESSAGE_TEXT "${KOKKOS_MESSAGE_TEXT}\n    GCC        5.3.0 or higher")
+SET(KOKKOS_MESSAGE_TEXT "${KOKKOS_MESSAGE_TEXT}\n    Intel     17.0.0 or higher")
+SET(KOKKOS_MESSAGE_TEXT "${KOKKOS_MESSAGE_TEXT}\n    NVCC      9.2.88 or higher")
 SET(KOKKOS_MESSAGE_TEXT "${KOKKOS_MESSAGE_TEXT}\n    HIPCC      3.5.0 or higher")
-SET(KOKKOS_MESSAGE_TEXT "${KOKKOS_MESSAGE_TEXT}\n    PGI         17.1 or higher\n")
+SET(KOKKOS_MESSAGE_TEXT "${KOKKOS_MESSAGE_TEXT}\n    PGI         17.4 or higher\n")
```
Comment (masterleinad) [inline cmake/kokkos_compiler_id.cmake:100]: Changing this file should probably not be part of this pull request. In fact, the changes should already be in `develop`. Would you mind rebasing?

**Comment by dhollman** [comment]
FYI this should be cleaned up, rebased, and ready for review.

**Comment by dalg24** [comment]
You need to fix the format

File: core/src/impl/Kokkos_Default_GraphNodeKernel.hpp [hunk 2]
```diff
@@ -63,7 +63,7 @@ template <class ExecutionSpace>
 struct GraphNodeKernelDefaultImpl {
   // TODO @graphs decide if this should use vtable or intrusive erasure via
   //      function pointers like in the rest of the graph interface
-  virtual void execute_kernel() const = 0;
```
Comment (crtrott) [inline core/src/impl/Kokkos_Default_GraphNodeKernel.hpp:66]: why does the constness change here and below?

**Review by crtrott — CHANGES_REQUESTED** [review]
Looks good generally, just need an explanation for the change of constness in some functions.

File: core/src/Cuda/Kokkos_Cuda_KernelLaunch.hpp [hunk 3]
```diff
@@ -356,6 +397,55 @@ struct CudaParallelLaunchKernelInvoker<
...
+
+      // Unlike in the non-graph case, we can get away with doing an async copy
+      // here because the `DriverType` instance is held in the GraphNodeImpl
+      // which is guaranteed to be alive until the graph instance itself is
+      // destroyed, where there should be a fence ensuring that the allocation
+      // associated with this kernel on the device side isn't deleted.
+      cudaMemcpyAsync(driver_ptr, &driver, sizeof(DriverType),
+                      cudaMemcpyDefault, cuda_instance->m_stream);
+
+      void const* args[] = {&driver_ptr};
```
Comment (dalg24) [inline core/src/Cuda/Kokkos_Cuda_KernelLaunch.hpp:426]: For reference, here is the diff with  `CudaParallelLaunchKernelInvoker<..., LocalMemory>:: create_parallel_launch_graph_node()`
```diff
--- local_memory
+++ global_memory
@@ -13,7 +13,17 @@
       Impl::check_shmem_request(cuda_instance, shmem);
       Impl::configure_shmem_preference(base_t::get_kernel_func(), prefer_shmem);

-      void const* args[] = {&driver};
+      auto* driver_ptr = Impl::allocate_driver_storage_for_kernel(driver);
+
+      // Unlike in the non-graph case, we can get away with doing an async copy
+      // here because the `DriverType` instance is held in the GraphNodeImpl
+      // which is guaranteed to be alive until the graph instance itself is
+      // destroyed, where there should be a fence ensuring that the allocation
+      // associated with this kernel on the device side isn't deleted.
+      cudaMemcpyAsync(driver_ptr, &driver, sizeof(DriverType),
+                      cudaMemcpyDefault, cuda_instance->m_stream);
+
+      void const* args[] = {&driver_ptr};

       cudaKernelNodeParams params = {};

File: core/src/impl/Kokkos_Default_GraphNodeKernel.hpp (hunk 2 above)
Comment (dhollman) [reply core/src/impl/Kokkos_Default_GraphNodeKernel.hpp:66]: Because `Impl::ParallelReduce<>::execute()` isn't `const` for the `Cuda` specialization, and we need to call that in `Cuda` versions before 10.0.

File: core/src/Cuda/Kokkos_Cuda_KernelLaunch.hpp (hunk 3 above)
Comment (dhollman) [reply core/src/Cuda/Kokkos_Cuda_KernelLaunch.hpp:426]: yeah, I should probably find a way to combine those two, sorry. I can clean it up later, but I tried to repeat as little code as possible

File: core/src/Cuda/Kokkos_Cuda_Parallel.hpp [hunk 4]
```diff
@@ -522,7 +522,8 @@ class ParallelFor<FunctorType, Kokkos::RangePolicy<Traits...>, Kokkos::Cuda> {
 template <class FunctorType, class... Traits>
 class ParallelFor<FunctorType, Kokkos::MDRangePolicy<Traits...>, Kokkos::Cuda> {
  public:
-  using Policy = Kokkos::MDRangePolicy<Traits...>;
+  using Policy       = Kokkos::MDRangePolicy<Traits...>;
+  using functor_type = FunctorType;
```
Comment (dalg24) [inline core/src/Cuda/Kokkos_Cuda_Parallel.hpp:526]: I was about to complain it is not consistent but it is already a mess in here...

File: core/src/Cuda/Kokkos_Cuda_KernelLaunch.hpp [hunk 5]
```diff
@@ -303,6 +305,45 @@ struct CudaParallelLaunchKernelInvoker<
          get_kernel_func())<<<grid, block, shmem, cuda_instance->m_stream>>>(
         driver);
   }
+
+#ifdef KOKKOS_CUDA_ENABLE_GRAPHS
+  inline static void create_parallel_launch_graph_node(
```
[the remark this answers is not in the document]
Comment (dhollman) [reply core/src/Cuda/Kokkos_Cuda_KernelLaunch.hpp:310]: No, because it will need to do more than just that in certain cases, and I don't think I should create separate customization points for the other things that may need to happen.

File: core/src/impl/Kokkos_Default_GraphNodeKernel.hpp (hunk 2 above)
Comment (crtrott) [reply core/src/impl/Kokkos_Default_GraphNodeKernel.hpp:66]: We looked at this a bit more. There are deeper issues here, and we may change the const ness back if we can resolve the deeper issues. Largely it has to do with the buffers for reductions which need to potentially grow, and the pointers which are part of the driver and hence have to change during execute. There are issues here with respect to multiple streams, mixing of delayed graph submit and non-graph dispatch, as well as multi threaded dispatch with the non-graph stuff. But we should go ahead with this PR and solve those corner cases separately.

**Review by crtrott — APPROVED** [review]

**Comment by dalg24** [comment]
@dhollman Please address my comments

**Review by dalg24 — CHANGES_REQUESTED** [review]
Please address my previous comments

File: core/src/Cuda/Kokkos_Cuda_GraphNodeKernel.hpp [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+#include <Kokkos_Parallel_Reduce.hpp>
+#include <Kokkos_PointerOwnership.hpp>
+
+#include <Kokkos_Cuda.hpp>
+#include <cuda_runtime_api.h>
+
+namespace Kokkos {
+namespace Impl {
+
+struct CudaGraphNodeKernelBase {};
```
[the remark this answers is not in the document]
Comment (dhollman) [reply core/src/Cuda/Kokkos_Cuda_GraphNodeKernel.hpp]: You're right; I don't need it anymore. Resolved by 778dc45a

File: core/src/Cuda/Kokkos_Cuda_GraphNode_Impl.hpp [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+      delete;
+
+  GraphNodeBackendSpecificDetails(GraphNodeBackendSpecificDetails&&) noexcept =
+      delete;
+
+  GraphNodeBackendSpecificDetails& operator   =(
+      GraphNodeBackendSpecificDetails const&) = delete;
+
+  GraphNodeBackendSpecificDetails& operator       =(
+      GraphNodeBackendSpecificDetails&&) noexcept = delete;
```
[the remark this answers is not in the document]
Comment (dhollman) [reply core/src/Cuda/Kokkos_Cuda_GraphNode_Impl.hpp]: Addressed in f985b368

File: core/src/Cuda/Kokkos_Cuda_GraphNode_Impl.hpp [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+template <>
+struct GraphNodeBackendSpecificDetails<Kokkos::Cuda> {
+  cudaGraphNode_t node = nullptr;
+
+  //----------------------------------------------------------------------------
+  // <editor-fold desc="Ctors, destructor, and assignment"> {{{2
+
+  explicit GraphNodeBackendSpecificDetails() = default;
+
+  GraphNodeBackendSpecificDetails(_graph_node_is_root_ctor_tag) noexcept {}
```
[the remark this answers is not in the document]
Comment (dhollman) [reply core/src/Cuda/Kokkos_Cuda_GraphNode_Impl.hpp]: Resolved by fd08cb3b

File: core/src/Cuda/Kokkos_Cuda_GraphNodeKernel.hpp [hunk 9]
```diff
@@ -0,0 +1,214 @@
...
+  }
+  void set_cuda_graph_node_ptr(cudaGraphNode_t* arg_node_ptr) {
+    m_graph_node_ptr = arg_node_ptr;
+  }
+  cudaGraphNode_t* get_cuda_graph_node_ptr() const { return m_graph_node_ptr; }
+  cudaGraph_t const* get_cuda_graph_ptr() const { return m_graph_ptr; }
+
+  Kokkos::ObservingRawPtr<base_t> allocate_driver_memory_buffer() const {
+    KOKKOS_EXPECTS(m_driver_storage == nullptr)
+    using record_t = typename GraphNodeKernelImpl::Record;
```
[the remark this answers is not in the document]
Comment (dhollman) [reply core/src/Cuda/Kokkos_Cuda_GraphNodeKernel.hpp]: Resolved by 1b4b277c

File: core/src/Cuda/Kokkos_Cuda_Parallel.hpp (hunk 4 above)
Comment (dhollman) [reply core/src/Cuda/Kokkos_Cuda_Parallel.hpp:526]: yup.

**Comment by dhollman** [comment]
@dalg24 should be done now :-)
